compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs

```diff
@@ -564,7 +564,12 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
         if !errors.is_empty() {
             self.adjust_fulfillment_errors_for_expr_obligation(&mut errors);
+            let errors_causecode = errors
+                .iter()
+                .map(|e| (e.obligation.cause.span, e.root_obligation.cause.code().clone()))
+                .collect::<Vec<_>>();
             self.err_ctxt().report_fulfillment_errors(errors);
+            self.collect_unused_stmts_for_coerce_return_ty(errors_causecode);
```

Comment:
@estebank 
I fixed the conflicts, please have a review.

this changed because of this commit: 
https://github.com/rust-lang/rust/commit/91b9ffeab0f8c491c81cd976c0d23d947cd6d12f

I need to clone `span` and `cause code` from errors for `collect_unused_stmts_for_coerce_return_ty`.

Any better solution?

Reply:
Ah, I see what's going on, `report_fulfillment_errors` now takes ownership, a change I made purely because we never needed access to the errors after reporting before this change.

Would it be feasible to instead modify the error cause code to carry an `Option<Span>` for the place where the `return` would be suggested? That way we don't need to leverage the "steal" mechanism nor clone the errors only iterate over them. We do something similar for errors coming from type parameters or from function arguments.

Reply:
Do you mean to add a field `Option<span>` to `ExprBindingObligation`, then use `ObligationCause`'s `map_code` API to update the trigger span?

I tried this solution and seems involve a wider change, and end with that we can not get the `diag` here if we don't use the "steal" mechanism:

https://github.com/rust-lang/rust/blob/25d38c48c3c7d06c14ae65e964a114b89c09756e/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs#L1889